CONTRIBUTING.md

```diff
@@ -0,0 +1,219 @@
+We discovered that secrets generally fit into one of 3 categories:
+
+1. Those that follow a pattern and contain identifiable sequences. For example, SonarQube API tokens always start with a known prefix (sqa_, sqp_ or squ_) that’s followed by 40 hexadecimal characters.
+2. Those that follow a pattern, but which can only be identified by looking at the surrounding text. For example, AWS secret keys are always 40 base64 characters long, but only the surrounding text can identify whether they’re used with AWS.
+3. Those that follow no fixed pattern, and which are only identifiable based on the surrounding text. For example, a PostgreSQL database password is user-chosen and must be set into a named environment variable before calling the psql command-line client.
+
+As you go down the list, the secrets get harder to detect without also raising false positives. Category #3 is particularly difficult because the suspected secret could be a dummy value (e.g. `[password]`) or a variable that’s replaced at runtime.
+
+A lot of tuning is needed in order to identify and exclude values that are common false positives.  
+  
+## Code structure for secret detection
+
+Take a look at the example of a secret detection specification file for a fake
+cloud provider below. This is considered the anatomy of a "good" secret detection file.
+
+``` yaml
+provider:
+  metadata:
+    name: FakeCloud
+    category: Cloud Provider
+    message: Make sure this FakeCloud API key gets revoked, changed, and removed from the code.
+  detection:
+    pre:
+      include:
+        content:
+          - "(?i)fakecloud"
+    post:
+      # Avoid matching values found on SourceGraph that look like dummy
+      # passwords or insertions like:
+      patternNot: 
+        - "(\\w)\\1{6,}"
+        - "(?i)rightcloud"
+        - "(?i)testkey|(s|ex)ample""
+  rules:
+    # The RSpec Key corresponds to the contents of SonarSource/rspec.
+    # To create a new secret, you need to create a new rule in rspec, which
+    # will output a new rule ID (rspec key)
+    - rspecKey: SXXXX
+      id: fakecloud-api-key
+      metadata:
+        # This does not influence the rule name
+        name: FakeCloud API Key
+
+      detection:
+        matching:
+          # Matching guidelines: As a best practice, make the regexes as less
+          # greedy as possible.
+          # To do so, start to understand how many characters are in each part
+          # of the secret. Avoid `+` or `*` for example.
+          pattern: "(?i)\\b(fakecloud-\\w{32})\\b"
+
+      # Add the compliant and non compliant example from the rspec as test cases
+      # to make sure that they are detected correctly
+      examples:
+        - text: |
+            SECRET=fakecloud-W1QrzMjSp5hYtO0KEvIU3C2dlx4GAfbn
+          containsSecret: true
+          match: fakecloud-W1QrzMjSp5hYtO0KEvIU3C2dlx4GAfbn
+        - text: |
+            BLABLA
+          containsSecret: false
+```
+
+## RSpec writing
```

Comment:
(This may be better kept in RSpec docs and linked here, but I don't know if this is allowed.)